File: 15-3sum/3sum.js
const threeSum = function (nums) {
    nums.sort((a, b) => a - b);
    const result = [];

    for (let i = 0; i < nums.length - 2; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) continue;

        let low = i + 1;
        let high = nums.length - 1;

        while (low < high) {
            const sum = nums[i] + nums[low] + nums[high];

            if (sum > 0) {
                high--;
            } else if (sum < 0) {
                low++;
            } else {

                result.push([nums[i], nums[low], nums[high]]);
                low++;
                high--;

                while (low < high && nums[low] === nums[low - 1]) {
                    low++;
                }
                while (low < high && nums[high] === nums[high + 1]) {
                    high--;
                }
            }
        }
    }

    return result;
};

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        result = []
        for index in range(n - 2):
            if index > 0 and nums[index] == nums[index - 1]:
                continue
            low = index + 1
            high = n - 1
            target = -1 * nums[index]
            while low < high:
                twoSum = nums[low] + nums[high]
                if twoSum == target:
                    result.append([nums[index], nums[low], nums[high]])
                    low += 1
                    high -= 1
                    while low < n and nums[low] == nums[low - 1]:
                        low += 1
                    while high >= 0 and nums[high] == nums[high + 1]:
                        high -= 1
                elif twoSum < target:
                    low += 1
                else:
                    high -= 1

        return result
